src/lib/tecs/TECS.cpp

```diff
@@ -41,241 +46,403 @@ using math::constrain;
+	 * The weighting can be adjusted between 0 and 2 depending on speed and altitude accuracy requirements.
+	*/
+	// Calculate the specific energy balance rate demand
+	float seb_rate_setpoint = se.spe.rate_setpoint * weight.spe_weighting - se.ske.rate_setpoint * weight.ske_weighting;
+
+	// Calculate the specific energy balance rate error
+	float seb_rate_error = (se.spe.rate_error * weight.spe_weighting) - (se.ske.rate_error * weight.ske_weighting);
+
+	_debug_output.energy_balance_rate_error = seb_rate_error;
+	_debug_output.energy_balance_rate_sp = seb_rate_setpoint;
+
+	if (param.integrator_gain_pitch > 0.0f) {
+		// Calculate pitch integrator input term
+		float pitch_integ_input = seb_rate_error * param.integrator_gain_pitch;
+
+		// Prevent the integrator changing in a direction that will increase pitch demand saturation
+		if (_pitch_setpoint > param.pitch_max) {
+			pitch_integ_input = min(pitch_integ_input, 0.f);
+
+		} else if (_pitch_setpoint < param.pitch_min) {
+			pitch_integ_input = max(pitch_integ_input, 0.f);
+		}
+
+		// Update the pitch integrator state.
+		_pitch_integ_state = _pitch_integ_state + pitch_integ_input * dt;
+
+	} else {
+		_pitch_integ_state = 0.0f;
+	}
+
+	// Calculate a specific energy correction that doesn't include the integrator contribution
+	float SEB_rate_correction = seb_rate_error * param.pitch_damping_gain + _pitch_integ_state + param.seb_rate_ff *
+				    seb_rate_setpoint;
+
+	// Calculate derivative from change in climb angle to rate of change of specific energy balance
+	const float climb_angle_to_SEB_rate = input.tas * CONSTANTS_ONE_G;
+
+	// During climbout, bias the demanded pitch angle so that a zero speed error produces a pitch angle
+	// demand equal to the minimum pitch angle set by the mission plan. This prevents the integrator
+	// having to catch up before the nose can be raised to reduce excess speed during climbout.
+	if (flag.climbout_mode_active) {
+		SEB_rate_correction += param.pitch_min * climb_angle_to_SEB_rate;
+	}
+
+	// Convert the specific energy balance rate correction to a target pitch angle. This calculation assumes:
+	// a) The climb angle follows pitch angle with a lag that is small enough not to destabilise the control loop.
+	// b) The offset between climb angle and pitch angle (angle of attack) is constant, excluding the effect of
+	// pitch transients due to control action or turbulence.
+	float pitch_setpoint_unc = SEB_rate_correction / climb_angle_to_SEB_rate;
+
+	float pitch_setpoint = constrain(pitch_setpoint_unc, param.pitch_min, param.pitch_max);
+
+	// Comply with the specified vertical acceleration limit by applying a pitch rate limit
+	// NOTE: at zero airspeed, the pitch increment is unbounded
+	const float pitch_increment = dt * param.vert_accel_limit / input.tas;
+	_pitch_setpoint = constrain(pitch_setpoint, _pitch_setpoint - pitch_increment,
```

Comment:
@KonradRudin Should we be consistent with the order of constraints with both throttle and pitch?